File: Practica 8/Problema2/Kotlin/AbstractAnimal.kt
import LifeForm

abstract class AbstractAnimal : LifeForm {
    companion object {
        private var maxId = 0
    }

    private var hunger = 0
    private val name: String
    private val id: Int
    private val year: Int

    constructor(name: String, year: Int, hunger: Int) {
        maxId++
        this.id = maxId
        this.hunger = hunger
        this.year = year
        this.name = name
    }

    constructor(name: String, year: Int) : this(name, year, 0)
    
    override fun eat(quantity: Int) {
        hunger -= quantity
        hunger = if (hunger < 0) 0 else hunger
    }

    override fun getHunger(): Int {
        return hunger
    }

    override fun getYear(): Int {
        return year
    }

    override fun getName(): String {
        return name
    }
    
    abstract override fun move(): String
    abstract override fun breathe(): String
    abstract override fun reproduce(): String
}

File: Practica 8/Problema2/Dart/AbstractAnimal.dart
import 'LifeForm.dart';

abstract class AbstractAnimal implements LifeForm {
  static int maxId = 0;
  late int hunger;
  late String name;
  late int id;
  late int year;

  AbstractAnimal(String name, int year, int hunger) {
    this.hunger = hunger;
    this.name = name;
    this.year = year;
    maxId++;
    id = maxId;
  }

  AbstractAnimal.namedYear(String name, int year) : this(name, year, 0);

  void eat(int quantity) {
    hunger -= quantity;
    hunger = hunger < 0 ? 0 : hunger;
  }

  String getName() {
    return name;
  }

  int getHunger() {
    return hunger;
  }

  int getId() {
    return id;
  }

  int getYear() {
    return year;
  }

  String move();
  String breathe();
  String reproduce();
}
